### ISSUE TITLE
The training is not coming out well, so I hope you can see it and respond!

### ISSUE THREAD

**Issue by Worker-Hou** [JuliaWolleb/Diffusion-based-Segmentation#45]
Hello! You are great at this job! But when I reproduce this code of yours, the segmentation result I get is very bad.
Below is the segmented image I used and the resulting image I got. The training set I used was the BraTS2020 training set using the 80th to 129th axial slices from 332 patients, and the test set was a selected set of images from 37 patients. The steps of the training are 80,000 steps.The final threshold was set to 0.5 as per your paper. not sure if it is because of the dataset. Can you provide me with the number of the patients in the training set you used so that I can train with good results, thanks again!

![1](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/b1426787-b4fb-4626-afc0-07d31af8f0c5)
![2](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/9fd09b07-fca8-4071-9cd1-b595be2009d1)
![3](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/26343875-d40d-43ec-978e-71938d666c42)
![4](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/d187969d-b170-4e80-baa8-d1b611eb57ec)
![5](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/3f880bd6-56e6-476f-8eca-b93632228ba5)
![6](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/19acd4a9-f3a2-43d5-bc95-3ef7be37537b)

**Comment by Worker-Hou** [comment]
Hopefully someone who knows how to fix this can help me. Thanks!

**Comment by Worker-Hou** [comment]
When I increase the training steps to 100000 steps, and test the 5 segmented images thus generated, there is always one or two images that work very poorly, approximating pure white noise or unable to find the place to be segmented, but the other three or four images are able to get better segmentation results instead. Does anyone have any idea why this is? Thank you very much!

**Comment by JuliaWolleb** [comment]
Good morning

I think he example images you sent look very weird. Only the last one looks as expected.
I don't really onderstand why you have 5 segmented images. Out of the 4 MR input channels, you should only get one output image. Did you implement this accordingly?

**Comment by Worker-Hou** [comment]
> Good morning
> 
> I think he example images you sent look very weird. Only the last one looks as expected. I don't really onderstand why you have 5 segmented images. Out of the 4 MR input channels, you should only get one output image. Did you implement this accordingly?

I am basing this on one of your papers:
"For every image of the test set, we sample 5 different segmentation masks. This implicitly defines an ensemble by averaging over the 5 masks and thresholding it at 0.5."
Then in the test phase we have it generate 5 different images. However, based on my current code reproduction, the segmentation results and the corresponding metrics are significantly improved when taking the number of training sessions up to 300,000, with the exception of some images that get poor results. This makes me wonder, in your paper you got good segmentation results after training only up to 60,000 steps, whereas I need quite long steps.
Thanks for your reply! Thanks again for your work and looking forward to what you do next!

**Comment by 1362303119** [comment]
@Worker-Hou Hello, I am new to diffusion modeling and have tried to reproduce the code, but I am having problems processing the BRATS2020 dataset. Can you share the code you used to process the dataset? My email is mailto:zhanglili_955@163.com，If you can help me to solve this problem, I would be very grateful!

**Comment by Worker-Hou** [comment]
> @Worker-Hou Hello, I am new to diffusion modeling and have tried to reproduce the code, but I am having problems processing the BRATS2020 dataset. Can you share the code you used to process the dataset? My email is [mailto:zhanglili_955@163.com](mailto:zhanglili_955@163.com)，If you can help me to solve this problem, I would be very grateful!

Hi, Ask ChatGpt this question it will give you the complete code!

**Comment by Worker-Hou** [comment]
> > @Worker-Hou Hello, I am new to diffusion modeling and have tried to reproduce the code, but I am having problems processing the BRATS2020 dataset. Can you share the code you used to process the dataset? My email is [mailto:zhanglili_955@163.com](mailto:zhanglili_955@163.com)，If you can help me to solve this problem, I would be very grateful!
> 
> Hi, Ask ChatGpt this question it will give you the complete code!

Have emailed you the code I used, FYI.

**Comment by 648lsf** [comment]
> Hello! You are great at this job! But when I reproduce this code of yours, the segmentation result I get is very bad. Below is the segmented image I used and the resulting image I got. The training set I used was the BraTS2020 training set using the 80th to 129th axial slices from 332 patients, and the test set was a selected set of images from 37 patients. The steps of the training are 80,000 steps.The final threshold was set to 0.5 as per your paper. not sure if it is because of the dataset. Can you provide me with the number of the patients in the training set you used so that I can train with good results, thanks again!
> 
> ![1](https://user-images.githubusercontent.com/128661908/264380909-b1426787-b4fb-4626-afc0-07d31af8f0c5.png) ![2](https://user-images.githubusercontent.com/128661908/264380937-9fd09b07-fca8-4071-9cd1-b595be2009d1.png) ![3](https://user-images.githubusercontent.com/128661908/264380983-26343875-d40d-43ec-978e-71938d666c42.png) ![4](https://user-images.githubusercontent.com/128661908/264381217-d187969d-b170-4e80-baa8-d1b611eb57ec.png) ![5](https://user-images.githubusercontent.com/128661908/264381276-3f880bd6-56e6-476f-8eca-b93632228ba5.png) ![6](https://user-images.githubusercontent.com/128661908/264381315-19acd4a9-f3a2-43d5-bc95-3ef7be37537b.png)

Hello! Can the results of your final experiment achieve the results in the paper? Why is it so difficult for me to achieve it?

**Comment by JuliaWolleb** [comment]
Did you double-check the input to your model? the first 4 channels should show the clean (i.e., non-noisy MR images). The 5th channel should show the noisy segmentation mask x_t. The loss is only computed on the noise of the 5th channel. The output should be a 1-channel (if you set`learn_sigma=False`) image.

**Comment by Worker-Hou** [comment]
> Did you double-check the input to your model? the first 4 channels should show the clean (i.e., non-noisy MR images). The 5th channel should show the noisy segmentation mask x_t. The loss is only computed on the noise of the 5th channel. The output should be a 1-channel (if you set) image.`learn_sigma=False`

Yes, my input image looks like the one below:
![image](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/f4d3af15-7c98-4ebf-9ab2-4acbb657a15e)
But your last hyperparameter setting: learn_sigma=False, makes me wonder, isn't your hyperparameter setting set to True?

**Comment by Worker-Hou** [comment]
> > 你好！你很擅长这份工作！但是当我重现你的这段代码时，我得到的分割结果非常糟糕。下面是我使用的分割图像和我得到的结果图像。我使用的训练集是 BraTS2020 训练集，使用来自 80 名患者的第 129 至 332 个轴向切片，测试集是来自 37 名患者的一组精选图像。培训的步骤是80,000步。根据您的论文，最终阈值设置为 0.5。不确定是否是因为数据集。你能不能给我提供你使用的训练集中的病人数量，这样我就可以训练出好的结果，再次感谢！
> > ![1](https://user-images.githubusercontent.com/128661908/264380909-b1426787-b4fb-4626-afc0-07d31af8f0c5.png) ![2](https://user-images.githubusercontent.com/128661908/264380937-9fd09b07-fca8-4071-9cd1-b595be2009d1.png) ![3](https://user-images.githubusercontent.com/128661908/264380983-26343875-d40d-43ec-978e-71938d666c42.png) ![4](https://user-images.githubusercontent.com/128661908/264381217-d187969d-b170-4e80-baa8-d1b611eb57ec.png) ![5](https://user-images.githubusercontent.com/128661908/264381276-3f880bd6-56e6-476f-8eca-b93632228ba5.png) ![6](https://user-images.githubusercontent.com/128661908/264381315-19acd4a9-f3a2-43d5-bc95-3ef7be37537b.png)
> 
> 你好！你最终的实验结果能达到论文中的结果吗？为什么我很难实现它？
Nor could I have achieved such good results using only a few steps in the author's paper. Could it be due to differences in the hardware equipment of each experiment? You can extend the training steps for as long as possible (say 200,000 steps or more).

**Comment by Worker-Hou** [comment]
> > 您是否仔细检查了模型的输入？前 4 个通道应显示干净（即无噪声的 MR 图像）。第 5 个通道应显示嘈杂的分段掩码x_t。损耗仅根据第 5 个通道的噪声计算。输出应为 1 通道（如果设置）图像。`learn_sigma=False`
> 
> 是的，我的输入图像如下所示： 但是您的最后一个超参数设置： ![image](https://user-images.githubusercontent.com/128661908/279308145-f4d3af15-7c98-4ebf-9ab2-4acbb657a15e.png) learn_sigma=False，让我想知道，您的超参数设置不是设置为 True 吗？

And when I look at the code, either this --learn_sigma=True or False -- doesn't affect the code.
![image](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/be54b4c9-d6ef-4c90-bf4d-66191b08405d)
![image](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/17628e1b-a74a-4287-9e64-81e988396fc0)
![image](https://github.com/JuliaWolleb/Diffusion-based-Segmentation/assets/128661908/1fa5b072-6e97-4b56-b23e-6546ba51ebb7)

**Comment by 648lsf** [comment]
我在3090上把batch_size设置为4,跑出来的模型到60000step,会有稍微变好,但是很难达到论文中的那样准确,原因可能是集成方式的问题

**Comment by JuliaWolleb** [comment]
Regarding `learn_sigma` flag:
If you set `learn_sigma=True`, you need 2 output channels: one for the mean, and one for the variance.
If you set `learn_sigma=False`, you need 1 output for the mean, and the variance is fixed.
Both work well, you just need to decide in the beginning, since this affects the architecture of the U-Net model.

@Worker-Hou , could you try to set the noise level to a small number on your _img input4_, such that we can check whether there is a binary segmentation mask underneath?

**Comment by Worker-Hou** [comment]
> Regarding `learn_sigma` flag: If you set `learn_sigma=True`, you need 2 output channels: one for the mean, and one for the variance. If you set `learn_sigma=False`, you need 1 output for the mean, and the variance is fixed. Both work well, you just need to decide in the beginning, since this affects the architecture of the U-Net model.
> 
> @Worker-Hou , could you try to set the noise level to a small number on your _img input4_, such that we can check whether there is a binary segmentation mask underneath?

Thanks for your reply, now I set learn_sigma=True for training, but no matter how I conduct training, it always fails to get good results if the training steps are less than 100,000 steps, I wonder if it is due to the problems of GPU and data set division. So far I've trained it to 300,000 steps to get relatively good results.

**Comment by dddb11** [comment]
@Worker-Hou Hello, I've encountered the same issue. I've noticed that the loss is decreasing quite rapidly in my training, but the performance is poor. So, I'd like to ask, is your loss consistently decreasing over 100,000 iterations, or is it relatively low from the beginning?

**Comment by JuliaWolleb** [comment]
In my experience, you can't rely on the loss function to monitor the training. The reason for this is that the amount of noise on the input images is random, so the loss starts to jump up and down. However, this does not mean that the model is not learning. If you want to check the segmentation performance, you need to evaluate some images of the validation set after each epoch.
The number of iterations needed also depends on your dataset, and how well the thresholding performs.

**Comment by dddb11** [comment]
> In my experience, you can't rely on the loss function to monitor the training. The reason for this is that the amount of noise on the input images is random, so the loss starts to jump up and down. However, this does not mean that the model is not learning. If you want to check the segmentation performance, you need to evaluate some images of the validation set after each epoch. The number of iterations needed also depends on your dataset, and how well the thresholding performs.

Thanks for your reply. The model indeed perform better after longer training.

**Comment by dddb11** [comment]
> In my experience, you can't rely on the loss function to monitor the training. The reason for this is that the amount of noise on the input images is random, so the loss starts to jump up and down. However, this does not mean that the model is not learning. If you want to check the segmentation performance, you need to evaluate some images of the validation set after each epoch. The number of iterations needed also depends on your dataset, and how well the thresholding performs.

Thanks for your reply.

**Comment by Worker-Hou** [comment]
> In my experience, you can't rely on the loss function to monitor the training. The reason for this is that the amount of noise on the input images is random, so the loss starts to jump up and down. However, this does not mean that the model is not learning. If you want to check the segmentation performance, you need to evaluate some images of the validation set after each epoch. The number of iterations needed also depends on your dataset, and how well the thresholding performs.

Thank you for your reply. I sampled the loss value of training by observing it before. Different Gpus may get different results after the same training steps. I have found a better result. Thanks again for your advice!

**Comment by Worker-Hou** [comment]
> @Worker-Hou 您好，我遇到了同样的问题。我注意到，在我的训练中，损失正在迅速减少，但表现很差。所以，我想问一下，你的损失是持续减少超过 100,000 次迭代，还是从一开始就相对较低？

正如作者说的，并不能完全通过观察损失值来选择所迭代的步骤，据我现在目前的实验来说，所迭代的步骤越大，往往最后的实验结果就越好，并且训练过程的损失值是随机变化的，因为使用的是L2损失函数。你可以适当的延长你的训练步骤来改善你的实验结果。
